### PR TITLE
try-runtime on_runtime_upgrade test

### PR DESCRIPTION
**Pull request by brunopgalvao**
This is a test for try-runtime's `on_runtime_upgrade`

This test will look for success otherwise it will fail and output errors.

Run the test:
```
cargo test --features try-runtime -p try-runtime-cli -- --nocapture
```

In order to replicate a bug that is pending this PR to be merged, this PR also includes the changes found here:
- https://github.com/paritytech/substrate/pull/13991

### CODE REVIEW

**Comment by liamaharon** [comment]
@brunopgalvao think we can also try to fix https://github.com/paritytech/substrate/issues/13969 in this PR?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2791806

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2808812

File: utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	use assert_cmd::cargo::cargo_bin;
+	use regex::Regex;
+	use std::{process, time::Duration, str::from_utf8};
+	use substrate_cli_test_utils as common;
+	use tokio::process::Command;
+
+	#[tokio::test]
+	async fn on_runtime_upgrade_works() {
+		// Build substrate so binaries used in the test use the latest code.
+		common::build_substrate(&["--features=try-runtime"]);
```
Comment (ggwpez) [inline utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs:32]: Why is this needed?  
The check above should already ensure that, or `feature = "try-runtime"`?  
I dont think we should need to build withing a test.

File: utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+#![cfg(unix)]
+
+#[cfg(feature = "try-runtime")]
+mod tests {
```
Comment (ggwpez) [inline utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs:22]:
```suggestion
#![cfg(all(unix, feature = "try-runtime"))]
```

File: utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs (hunk 1 above)
Comment (liamaharon) [reply utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs:32]: @ggwpez we build within the test because we spin up an actual substrate binary in the test

If we don't build prior to running, the binary used may not match the code on disk

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2817050

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2817049

File: utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs (hunk 1 above)
Comment (brunopgalvao) [reply utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs:31]: Removed the `#[cfg(feature = "try-runtime")]` feature flag as I do not see any difference, tests still work without it.

File: utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs (hunk 2 above)
Comment (brunopgalvao) [reply utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs]: `#[cfg(feature = "try-runtime")]` removed.

Comment (brunopgalvao) [reply utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs]: If there is a reason to keep this in there please let me know.

File: utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs (hunk 1 above)
Comment (ggwpez) [reply utils/frame/try-runtime/cli/tests/on_runtime_upgrade.rs:31]: Weird, since the [node CLI tests](https://github.com/paritytech/substrate/blob/a1d3d5004379976b13deab86b975c04ad78bcd72/bin/node/cli/tests/benchmark_block_works.rs#L33) do not need to re-build.  
Maybe the tests are in the wrong directory?
